core/shared/platform/windows/win_clock.c

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (C) 2023 Amazon Inc.  All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ */
+
+#include "platform_api_vmcore.h"
+#include <winternl.h>
+
+#define NANOSECONDS_PER_SECOND 1000000000
+#define NANOSECONDS_PER_TICK 100
+
+static uint64
+calculate_monotonic_clock_frequency()
+{
+    LARGE_INTEGER frequency;
+    QueryPerformanceFrequency(&frequency);
+    return (uint64)frequency.QuadPart;
+}
+
+// The implementation below dervies from the following source:
+// https://github.com/WasmEdge/WasmEdge/blob/b70f48c42922ce5ee7730054b6ac0b1615176285/lib/host/wasi/win.h#L210
+static uint64
+from_file_time_to_wasi_timestamp(FILETIME filetime)
+{
+    static const uint64 ntto_unix_epoch =
+        134774 * 86400 * NANOSECONDS_PER_SECOND;
+
+    ULARGE_INTEGER temp = { .LowPart = filetime.dwLowDateTime,
+                            .HighPart = filetime.dwHighDateTime };
+
+    return temp.QuadPart - ntto_unix_epoch;
+}
+
+uint64
+current_value_of_peformance_counter()
+{
+    LARGE_INTEGER counter;
+    QueryPerformanceCounter(&counter);
+    return counter.QuadPart;
+}
+
+int
+os_clock_res_get(bh_clock_id_t clock_id, uint64 *resolution)
+{
+    switch (clock_id) {
+        case BH_CLOCK_ID_MONOTONIC:
+        {
+            const uint64 result = (uint64)NANOSECONDS_PER_SECOND
+                                  / calculate_monotonic_clock_frequency();
+            *resolution = result;
+            return BHT_OK;
+        }
+        case BH_CLOCK_ID_REALTIME:
+        case BH_CLOCK_ID_PROCESS_CPUTIME_ID:
+        case BH_CLOCK_ID_THREAD_CPUTIME_ID:
+        {
+            PULONG maximum_time;
+            PULONG minimum_time;
+            PULONG current_time;
+            NTSTATUS
+            status = NtQueryTimerResolution(&maximum_time, &minimum_time,
```

Comment:
This is an undocumented, internal API so it might be worth loading it at runtime (so we don't end up calling it if it's unavailable)? Also I'm not sure if it would be available on UWP.

Reply:
nvm, from what I can tell ntdll is always loaded by the process and even though this is undocumented it seems to be used quite ubiquitously anyway.